### PR TITLE
create ami-sets tool

### PR DESCRIPTION
**Pull request by anarute**
Add ami-sets tool that manages AMI sets from the AWS provisioner. 
It makes possible to add, edit and delete AMI sets.

### CODE REVIEW

**Comment by djmitche** [comment]
Just to set context for the review: this is intended to be functional enough to edit AMI sets as @anarute said.  There are UI improvements to be made (for example, a table of AMIs instead of a JSON blob), but given the schedule we're going to go back to focus on the provisioner backend before addressing those improvements.

**Comment by anarute** [comment]
Exactly. Just to point out for further improvements, we also need to add schema and data validations.

**Comment by jhford** [comment]
I can't get this to run locally.
```
mutating the [[Prototype]] of an object will cause your code to run very slowly; instead create the object with the correct initial [[Prototype]] value using Object.create common.bundle.js:12366:5
"Download the React DevTools for a better development experience: https://fb.me/react-devtools" common.bundle.js:54179:9
TypeError: undefined has no properties
require<[818]<()
 common.bundle.js:69622
s()
 common.bundle.js:1
s/<()
 common.bundle.js:1
require<[821]</<()
 common.bundle.js:70929
require<[821]<()
 common.bundle.js:70922
s()
 common.bundle.js:1
s/<()
 common.bundle.js:1
require<["taskcluster-client"]<()
 common.bundle.js:100254
s()
 common.bundle.js:1
s()
 app.bundle.js:1
s/<()
 app.bundle.js:1
[9]<()
 app.bundle.js:1110
s()
 app.bundle.js:1
s/<()
 app.bundle.js:1
[1]</<()
 app.bundle.js:7
[1]<()
 app.bundle.js:2
s()
 app.bundle.js:1
s/<()
 app.bundle.js:1
[2]<()
 app.bundle.js:354
s()
 app.bundle.js:1
s/<()
 app.bundle.js:1
[3]<()
 app.bundle.js:554
s()
 app.bundle.js:1
e()
 app.bundle.js:1
<anonymous>
 app.bundle.js:1
```

File: ami-sets/amiseteditor.jsx [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+  }
+};
+var decodeUserData = (obj) => {
+  if (obj && obj.UserData) {
+    obj.UserData = JSON.parse(new Buffer(obj.UserData, 'base64').toString());
+  }
+};
+
+var initialAmiSet ={
+  "amis": [{
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove quotes from property keys.

File: ami-sets/amiseteditor.jsx [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  },
+
+  getDefaultProps() {
+    return {
+      currentAmiSet:  ''     // '' implies. "Create AMI Set"
+    };
+  },
+
+  getInitialState() {
+
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove extra line.

File: ami-sets/amiseteditor.jsx [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    return {
+      // Loading amiSet or loaded amiSet
+      amiSetLoaded: false,
+      amiSetError:  undefined,
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Prefer `null` over `undefined`.

File: ami-sets/amiseteditor.jsx [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+      working:          false,
+      error:            null,
+      showAmiSet:       false
+    };
+  },
+
+  /** Load initial state */
+  load() {
+    // If there is no currentAmiSet, we're creating a new AMI Set
+    if (this.props.currentAmiSet === '') {
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```js
if (!this.props.currentAmiSet) {
```

File: ami-sets/amiseteditor.jsx [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        amiSet:           '',
+        amis:             initialAmiSet,
+        editing:          true,
+        working:          false,
+        error:            null
+      };
+    } else {
+      // Load currentAmiSet
+
+      //var amisObject = this.awsProvisioner.amiSet(this.props.currentAmiSet);
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove this comment.

File: ami-sets/amiseteditor.jsx [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+      return {
+        amiSet:           '',
+        amis:             initialAmiSet,
+        editing:          true,
+        working:          false,
+        error:            null
+      };
+    } else {
+      // Load currentAmiSet
+
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove extra line.

File: ami-sets/amiseteditor.jsx [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        editing:          false,
+        working:          false,
+        error:            null,
+        showAmiSet:       false
+      };
+    }
+  },
+
+  render() {
+
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove extra line.

File: ami-sets/amiseteditor.jsx [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    return (
+      <span>
+        {
+          this.props.currentAmiSet ? (
+            <div>
+              <h3>Update <code>{this.props.currentAmiSet}</code></h3>
+              {
+                isEditing ?(
+                  <div>
+                      {this.renderCodeEditor()}
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Indent this section using 2 spaces.

File: ami-sets/amiseteditor.jsx [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                      <br/>
+                      {this.renderEditingToolbar()}
+                    </div>
+                  ) : (
+                  <span>
+                    <pre>{ JSON.stringify(_.pick(this.state.amis, ['amis']), null, 2) }</pre>
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove spaces surrounding `JSON.stringify()`.

File: ami-sets/amiseteditor.jsx [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+                    <bs.ButtonToolbar>
+                      <bs.Button bsStyle="success"
+                        onClick={this.startEditing}>
+                        <bs.Glyphicon glyph="pencil"/>&nbsp;Edit AMI Set
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove `&nbsp;` and just insert a space.

File: ami-sets/amiseteditor.jsx [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+          ) : (
+            <div>
+            <bs.Input
+              type='text'
+              value={this.state.amiSet}
+              placeholder="amiSet"
+              label='AmiSet'
+              hasFeedback
+              ref='amiSet'
+              onChange={this.amiSetChange}/>
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Insert space before tag closing ` />`.

File: ami-sets/amiseteditor.jsx [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+  },
+
+  /** Render editing toolbar */
+  renderCodeEditor() {
+    return (
+      <CodeMirror
+        ref="amis"
+        lineNumbers={true}
+        mode="application/json"
+        textAreaClassName={'form-control'}
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```jsx
textAreaClassName="form-control"
```

File: ami-sets/amiseteditor.jsx [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        textAreaStyle={{minHeight: '20em'}}
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```jsx
textAreaStyle={{ minHeight: '20em' }}
```

File: ami-sets/amiseteditor.jsx [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        value={JSON.stringify(_.pick(this.state.amis, ['amis']), null, 2)}
+        onChange={this.onAmiSetChange}
+        indentWithTabs={true}
+        tabSize={2}
+        lint={true}
+        gutters={["CodeMirror-lint-markers"]}
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```jsx
gutters={['CodeMirror-lint-markers']}
```

File: ami-sets/amiseteditor.jsx [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        theme="ambiance"/>
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Insert space before closing tag `/>`.

File: ami-sets/amiseteditor.jsx [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+    );
+  },
+
+  /** Render editing toolbar */
+  renderEditingToolbar() {
+    return (
+      <bs.ButtonToolbar>
+        <bs.Button bsStyle="success"
+                   onClick={this.saveAmiSet}
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Indent using 2 spaces.

File: ami-sets/amiseteditor.jsx [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+                   disabled={this.state.working}>
+          <bs.Glyphicon glyph="ok"/>&nbsp;Save Changes
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Replace `&nbsp` with space.

File: ami-sets/amiseteditor.jsx [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+          success="AMI Set deleted">
+          Are you sure you want to delete AMI Set &nbsp;
+          <code>{this.state.amiSet}</code>?
+        </ConfirmAction>
+      </bs.ButtonToolbar>
+    );
+  },
+
+  startEditing() {
+    this.setState({editing: true});
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```js
this.setState({ editing: true });
```

**Comment by djmitche** [comment]
@jhford did you re-run `npm install`?  It looks like it's failing inside of `require("taskcluster-client")` based on that traceback..

File: ami-sets/amiseteditor.jsx [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+  },
+
+  onAmiSetChange(e) {
+    this.setState({amis: JSON.parse(e.target.value)});
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Use spaces to padding the interior of object literal braces:

```js
this.setState({ amis: JSON.parse(e.target.value) });
```

File: ami-sets/amiseteditor.jsx [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+
+  async saveAmiSet() {
+    try {
+      await this.awsProvisioner.updateAmiSet(this.state.amiSet, this.state.amis);
+      this.setState({
+        editing: false,
+        error:   null
+      });
+    } catch(err) {
+      this.setState({error: err});
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```js
this.setState({ error: err });
```

File: ami-sets/amiseteditor.jsx [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+    try {
+      await this.awsProvisioner.createAmiSet(this.state.amiSet, this.state.amis);
+      this.setState({
+        editing: false,
+        error:   null
+      });
+      this.props.selectAmiSet(this.state.amiSet);
+      this.props.refreshAmiSetsList();
+    } catch(err) {
+    this.setState({error: err});
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Indent 2 spaces.

Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```js
this.setState({ error: err });
```

File: ami-sets/amiseteditor.jsx [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+    }
+  },
+
+  async deleteAmiSet() {
+    await this.awsProvisioner.removeAmiSet(this.state.amiSet);
+    this.props.selectAmiSet(undefined);
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]:
```js
this.props.selectAmiSet();
```

File: ami-sets/amiseteditor.jsx [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+    this.props.refreshAmiSetsList();
+  }
+
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Remove extra line.

File: ami-sets/amisetmanager.jsx [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+      keys:                   ['selectedAmiSet'],
+      type:                   'string'
+    })
+  ],
+
+  /** Create an initial state */
+  getInitialState() {
+    return {
+      amiSetsLoaded:      false,
+      amiSetsError:       undefined,
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Use `null` over `undefined`.

File: ami-sets/amisetmanager.jsx [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+
+  /** Render user-interface */
+  render() {
+    return (
+      <bs.Row>
+        <bs.Col md={5}>
+          {this.renderAmiSetsTable()}
+          <bs.ButtonToolbar>
+            <bs.Button bsStyle="primary"
+                       onClick={this.selectAmiSet.bind(this, '')}
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Indent using 2 spaces.

```jsx
<bs.Button
  bsStyle="primary"
  onClick={this.selectAmiSet.bind(this, '')}
  disabled={this.state.selectedAmiSet === ''}>

```

File: ami-sets/amisetmanager.jsx [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+                       disabled={this.state.selectedAmiSet === ''}>
+              <bs.Glyphicon glyph="plus"/>
+              &nbsp;
+              Add AMI Set
+            </bs.Button>
+            <bs.Button bsStyle="success"
+                       onClick={this.reload}
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Indent using 2 spaces.

File: ami-sets/amisetmanager.jsx [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+                       disabled={!this.state.amiSetsLoaded}>
+              <bs.Glyphicon glyph="refresh"/>
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]:
```jsx
<bs.Glyphicon glyph="refresh"/> Refresh
```

File: ami-sets/amisetmanager.jsx [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+              &nbsp;
+              Refresh
+            </bs.Button>
+          </bs.ButtonToolbar>
+        </bs.Col>
+        <bs.Col md={7}>
+          <AmiSetEditor currentAmiSet={this.state.selectedAmiSet}
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Indent using 2 spaces.

File: ami-sets/amisetmanager.jsx [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+          {this.state.amiSets.map(this.renderAmiSetRow)}
+        </tbody>
+      </bs.Table>
+    );
+  },
+
+  /** Render row with amiSet */
+  renderAmiSetRow(amiSet, index) {
+    var isSelected = (this.state.selectedAmiSet === amiSet);
+    return (
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Insert line between last declaration and expressions.

File: ami-sets/amisetmanager.jsx [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+        <tbody>
+          {this.state.amiSets.map(this.renderAmiSetRow)}
+        </tbody>
+      </bs.Table>
+    );
+  },
+
+  /** Render row with amiSet */
+  renderAmiSetRow(amiSet, index) {
+    var isSelected = (this.state.selectedAmiSet === amiSet);
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]:
```js
var isSelected = this.state.selectedAmiSet === amiSet;
```

File: ami-sets/amisetmanager.jsx [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+    return (
+      <tr key={index}
+          className={isSelected ? 'info' : undefined}
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]:
```jsx
className={isSelected ? 'info' : ''}
```

Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Indent using 2 spaces.

File: ami-sets/amisetmanager.jsx [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+    );
+  },
+
+  async reloadAmiSet(amiSetId) {
+    // Load amiSet ignore errors (assume amiSet doesn't exist)
+    let amiSet = await this.awsProvisioner.amiSet(amiSetId).catch(() => null);
+    let selectedAmiSet = amiSetId;
+    let amiSets = _.cloneDeep(this.state.amiSets);
+    var index = _.findIndex(amiSets, a => a.amiSet === amiSetId);
+    if (index === -1 && amiSet !== null) {
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Insert line between last declaration and conditional.

File: ami-sets/amisetmanager.jsx [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+      amiSets.push(amiSet);
+    } else if (amiSet !== null) {
+      amiSets[index] = amiSet;
+    } else {
+      amiSets = amiSets.filter(aSet => aSet.amiSet !== amiSetId);
+    }
+    if (_.findIndex(amiSets, aSet => aSet.amiSet === selectedAmiSet) === -1) {
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Insert line between conditionals.

File: ami-sets/amisetmanager.jsx [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+      selectedAmiSet = '';
+    }
+    amiSets.sort((a, b) => a.amiSet > b.amiSet);
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]: Insert line between last conditional and next expression.

File: ami-sets/amisetmanager.jsx [hunk 35]
```diff
@@ -0,0 +1,144 @@
...
+    this.setState({amiSets, selectedAmiSet});
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]:
```js
this.setState({ amiSets, selectedAmiSet });
```

File: ami-sets/amisetmanager.jsx [hunk 36]
```diff
@@ -0,0 +1,144 @@
...
+  },
+
+  refreshAmiSetsList() {
+    this.reload();
+  },
+
+  selectAmiSet(amiSet) {
+    this.setState({
```
Comment (eliperelman) [inline ami-sets/amisetmanager.jsx]:
```js
this.setState({ selectedAmiSet: amiSet });
```

File: ami-sets/app.jsx [hunk 37]
```diff
@@ -0,0 +1,17 @@
...
+let utils         = require('../lib/utils');
+
+var hashManager = utils.createHashManager({
+  separator:            '/'
+});
+
+// Render component
+$(function() {
+  ReactDOM.render(
+    <AmiSetManager hashEntry={hashManager.root()}/>,
```
Comment (eliperelman) [inline ami-sets/app.jsx]: Insert space before tag closing.

File: ami-sets/app.less [hunk 38]
```diff
@@ -0,0 +1,9 @@
+@import "amiseteditor.less";
+
```
Comment (eliperelman) [inline ami-sets/app.less]: Remove extra line.

File: ami-sets/app.less [hunk 39]
```diff
@@ -0,0 +1,9 @@
...
+@import "./lib/format.less";
+
+.ami-set-manager-table {
+  tr{
```
Comment (eliperelman) [inline ami-sets/app.less]: Insert space before curly brace.

**Comment by eliperelman** [comment]
Gonna make 2 major-ish requests. In your changed lines, please remove all the extra spacing that lines up statements, e.g.:

```js
var React             = require('react');
var bs                = require('react-bootstrap');
var utils             = require('../lib/utils');
var taskcluster       = require('taskcluster-client');
var _                 = require('lodash');
var CodeMirror        = require('react-code-mirror');
var ConfirmAction     = require('../lib/ui/confirmaction');
```

or:

```js
  mixins: [
    utils.createTaskClusterMixin({
      clients: {
        awsProvisioner:       taskcluster.createClient(reference)
      },
      clientOpts: {
        awsProvisioner: {
          baseUrl:      'https://aws-provisioner.taskcluster.net/v1'
        }
      },
      reloadOnProps: ['currentAmiSet']
    })
  ]
```

Please change your modifications to just use normal spacing without lining up expressions.

Also, there are several places where comments are just words that are the same as the statement being mentioned, i.e. the comment is redundant and is not saying anything the code already tells you. Please remove these comments, or change them to be descriptive.

Thanks!

**Comment by anarute** [comment]
@eliperelman thanks for the review. I just have a question, I'm not sure if I got it right about "removing all the extra spacing that lines up statements". Should I change to this:

```
var React = require('react');
var bs = require('react-bootstrap');
```
instead of this?
```
var React       = require('react');
var bs          = require('react-bootstrap');
```

Also, when you say to change my modification to use normal spacing, what does it mean? is it the same thing I'm asking above?

**Comment by eliperelman** [comment]
@anarute yes, that is correct, those were the same request.

**Comment by jhford** [comment]
If I try to create a valid AMI Set without a name, I don't get an error, but rather "Are you sure that you would like tocreatethe AMI Set?".  If I confirm, I get a spinner and when the spinner finishes, i'm asked the question again.

If I name it 'hi', I get a similar outcome.  If I name it in the pattern of the existing ones, I get the same issue.  It seems that I don't get any error messages and I'm unable to actually create amisets.

**Comment by jhford** [comment]
```browser.js:50Warning: `<Input>` is deprecated. Use `<FormControl>`, `<Checkbox>`, or `<Radio>`, with `<FormGroup>` and/or `<InputGroup>` as needed instead.```

```browser.js:50Warning: "hasFeedback" property of "FormGroup" has been deprecated.
Use a `<FormControl.Feedback>` element.```

I'm not sure if this is coming from your code or one of our libraries.

**Comment by jhford** [comment]
It looks like we're getting a 404 for the case of not having a name.

```XMLHttpRequest cannot load https://aws-provisioner.taskcluster.net/v1/ami-set/. Response for preflight has invalid HTTP status code 404```

The issue when there is a name looks like it's related to me not being logged in with that browser.  The issue here, though, is that it should show me an error message instead of just restarting that flow.

**Comment by jhford** [comment]
Once I sign in, I also get this error:

```
warning.js:46 Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the ConfirmAction component.
```

It looks like the tool can create it, but we should deal with this exception and also the issue about the error message for the 404 issue.

**Comment by eliperelman** [comment]
@jhford 

```
browser.js:50Warning: `<Input>` is deprecated. Use `<FormControl>`, `<Checkbox>`, or `<Radio>`, with `<FormGroup>` and/or `<InputGroup>` as needed instead.

browser.js:50Warning: "hasFeedback" property of "FormGroup" has been deprecated. Use a `<FormControl.Feedback>` element.
```

This is from an update in a library we are using. These will eventually go away, but we are currently stuck with them.

File: ami-sets/amiseteditor.jsx [hunk 40]
```diff
@@ -0,0 +1,245 @@
...
+};
+var decodeUserData = (obj) => {
+  if (obj && obj.UserData) {
+    obj.UserData = JSON.parse(new Buffer(obj.UserData, 'base64').toString());
+  }
+};
+
+var initialAmiSet ={
+  amis: [{
+    region: "...",
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Single quotes.

File: ami-sets/amiseteditor.jsx [hunk 41]
```diff
@@ -0,0 +1,245 @@
...
+    obj.UserData = new Buffer(JSON.stringify(obj.UserData)).toString('base64');
+  }
+};
+var decodeUserData = (obj) => {
+  if (obj && obj.UserData) {
+    obj.UserData = JSON.parse(new Buffer(obj.UserData, 'base64').toString());
+  }
+};
+
+var initialAmiSet ={
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Space after equals sign.

File: ami-sets/amiseteditor.jsx [hunk 42]
```diff
@@ -0,0 +1,245 @@
...
+    let isCreating = isEditing && !this.props.currentAmiSet;
+
+    return (
+      <span>
+        {
+          this.props.currentAmiSet ? (
+            <div>
+              <h3>Update <code>{this.props.currentAmiSet}</code></h3>
+              {
+                isEditing ?(
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Space after question mark.

File: ami-sets/amiseteditor.jsx [hunk 43]
```diff
@@ -0,0 +1,245 @@
...
+                      </bs.Button>
+                    </bs.ButtonToolbar>
+                  </span>
+                )
+              }
+            </div>
+          ) : (
+            <div>
+              <bs.Input
+                type='text'
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Use double quotes for markup.

File: ami-sets/amiseteditor.jsx [hunk 44]
```diff
@@ -0,0 +1,245 @@
...
+                placeholder="amiSet"
+                label='AmiSet'
+                hasFeedback
+                ref='amiSet'
+                onChange={this.amiSetChange} />
+              {this.renderCodeEditor()}
+              <br/>
+              <bs.ButtonToolbar>
+                <ConfirmAction
+                  buttonStyle='primary'
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Double quotes for markup.

File: ami-sets/amiseteditor.jsx [hunk 45]
```diff
@@ -0,0 +1,245 @@
...
+                  glyph='ok'
+                  label={this.props.amiSet ? 'Update AmiSet' : 'Create AmiSet'}
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: This is fine. Use double quotes for markup attributes, but single quotes in the interpolated values.

File: ami-sets/amiseteditor.jsx [hunk 46]
```diff
@@ -0,0 +1,245 @@
...
+                  action={this.props.amiSet ? this.save : this.create}
+                  success='Saved AMI Set'>
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Double quotes for markup attributes.

File: ami-sets/amiseteditor.jsx [hunk 47]
```diff
@@ -0,0 +1,245 @@
...
+                  Are you sure that you would like to
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Indent this section 2 more spaces.

File: ami-sets/amiseteditor.jsx [hunk 48]
```diff
@@ -0,0 +1,245 @@
...
+    );
+  },
+
+  renderCodeEditor() {
+    return (
+      <CodeMirror
+        ref="amis"
+        lineNumbers={true}
+        mode="application/json"
+        textAreaClassName='form-control'
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Double quotes.

File: ami-sets/amiseteditor.jsx [hunk 49]
```diff
@@ -0,0 +1,245 @@
...
+    return (
+      <bs.ButtonToolbar>
+        <bs.Button
+          bsStyle="success"
+          onClick={this.saveAmiSet}
+          disabled={this.state.working}>
+          <bs.Glyphicon glyph="ok"/> Save Changes
+        </bs.Button>
+        <ConfirmAction
+          buttonStyle='danger'
```
Comment (eliperelman) [inline ami-sets/amiseteditor.jsx]: Double quote markup attributes.

**Pull request merged**